### PR TITLE
Support matchers in labels API

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

As pr https://github.com/prometheus/prometheus/pull/8301 was merged, this pr adds `matches` to `LabelNames` and `LabelValues` APIs. 

One thing I am not very sure is the testing, seems we are just mocking the HTTP response in the unit tests, it cannot really ensure correctness. But I already test this manually and it works.

### CODE REVIEW

File: api/prometheus/v1/api.go
Comment (bwplotka) [inline api/prometheus/v1/api.go:234]: This is technically a breaking change. I am not sure what are the rules for this project cc @beorn7 

Do we allow this or not in single major version?

**Review by bwplotka — COMMENTED** [review]
Hmm I think overall LGTM, except the compatibility break. Not sure what is the way forward here.

In terms of tests we test what we have to test, no? We check if parameters are passed correctly to the underlying URL query.

File: api/prometheus/v1/api.go
Comment (yeya24) [reply api/prometheus/v1/api.go:234]: > In terms of tests we test what we have to test, no? We check if parameters are passed correctly to the underlying URL query.

Now only HTTP method and path are tested https://github.com/prometheus/client_golang/blob/master/api/prometheus/v1/api_test.go#L71-L76. We don't check parameters.

**Comment by yeya24** [comment]
> In terms of tests we test what we have to test, no? We check if parameters are passed correctly to the underlying URL query.

Now only HTTP method and path are tested https://github.com/prometheus/client_golang/blob/master/api/prometheus/v1/api_test.go#L71-L76. We don't check parameters.

File: api/prometheus/v1/api.go
Comment (beorn7) [reply api/prometheus/v1/api.go:234]: The whole API client is marked as experimental and not covered by the semantic versioning.

(In fact, I'd much prefer to have the API client on a separate release schedule. But as long as the API client is effectively pre-1.0, it doesn't really matter. Once we are ready to declare the API client as stable, we can revisit the decision if we want to keep it bundled with the rest of client_golang or separate it out (into a new repo or a separate Go module in a sub-directory).)

File: api/prometheus/v1/api.go
Comment (beorn7) [inline api/prometheus/v1/api.go:233]: I think this doc comment should be adjusted (it apparently didn't get adjusted when the time range was added, and now we have even matchers).

**Review by beorn7 — COMMENTED** [review]
No worries about the breaking change. That's still OK with the API client.

@bwplotka I leave it to you to approve and merge this once you are satisfied.

File: api/prometheus/v1/api.go
Comment (yeya24) [reply api/prometheus/v1/api.go:233]: Got it. I will update this.

**Comment by jotak** [comment]
Hello @yeya24  @bwplotka , thank you for this
Do you think it can be merged before the next client release?

**Comment by yeya24** [comment]
Updated. Please take a look @beorn7 @bwplotka if you have time.

**Comment by beorn7** [comment]
Thanks @yeya24 . I have no other comments, but I'd much prefer if @bwplotka could give his final approval.

**Review by beorn7 — APPROVED** [review]
No word from @bwplotka . I'll merge this now.

**Comment by zhamdoctor** [comment]
> and path are tested

image
image
image
image
max(delta(go_memstats_alloc_bytes[5m])),this promql was transferred to unicode pattern,and causes an error.

**Pull request merged**
